### PR TITLE
bugfix/12716-axis-label-outside

### PR DESCRIPTION
**Pull request by wchmiel**
Fixed issue with axis labels outside plot range when ordinal enabled, see #12716.

### CODE REVIEW

File: ts/parts/Axis.ts [hunk 1]
```diff
@@ -5496,6 +5498,20 @@ extend(Axis.prototype, /** @lends Highcharts.Axis.prototype */{
             }
             if (!tickPositionsOption && !tickPositioner) {
                 this.adjustTickAmount();
+
+                // Make sure ticks are within axis min and max range (#12716).
+                if (this.isOrdinal) {
```
Comment (pawelfus) [inline ts/parts/Axis.ts]: Can we move this logic to `OrdinalAxis` file? E.g. add `afterAdjustTickAmount` event and modify config there.

File: ts/parts/Axis.ts [hunk 2]
```diff
@@ -5496,6 +5498,20 @@ extend(Axis.prototype, /** @lends Highcharts.Axis.prototype */{
...
+                    min = this.min;
+                    max = this.max;
+
+                    tickPositions = tickPositions.filter(
+                        (pos: number): boolean =>
+                            defined(min) && pos >= min &&
+                            defined(max) && pos <= max
```
Comment (pawelfus) [inline ts/parts/Axis.ts]: Instead of checking `defined(min)` in every iteration, can we don't call `filter` at all?

**Comment by pawelfus** [comment]
General question: Why these ticks and labels are rendered in the first place? Instead of filtering out ticks, we shouldn't create them at all. Not that tick is not always present there:

ticks

Seems like some threshold is set there, once crossed label hides/shows up (half of the range, e.g. half day?).

PS: Since it's a Highstock issue, please add `Product: Highstock` label

**Comment by wchmiel** [comment]
Additional tick is rendered because of this condition:
```js
while ((this.min as any) - minPointOffset > tickPositions[0]) {
    tickPositions.shift();
}
```

In some cases, this condition is false and tick is not removed.

**Comment by TorsteinHonsi** [comment]
Pending @pawelfus 's review

File: ts/parts/OrdinalAxis.ts [hunk 3]
```diff
@@ -1074,6 +1074,25 @@ addEvent(Axis, 'afterSetScale', function (this: Highcharts.Axis): void {
...
+
+// Make sure ticks are within axis min and max range (#12716).
+addEvent(Axis, 'afterTrimTicks', function (
+    this: Highcharts.Axis,
+    e: { tickPositions: Array<number> }
+): void {
+    var axis = this,
+        tickPositions = e.tickPositions;
+
+    if (axis.isOrdinal && axis.min && axis.max && tickPositions) {
```
Comment (pawelfus) [inline ts/parts/OrdinalAxis.ts]: Marginal case: this won't work with axis.min/max = 0

**Comment by pawelfus** [comment]
Interesting. In theory `minPointOffset` incorporates ordinalCorrection. Data is irregular (for one month number of milliseconds varies) so it takes an effect. @TorsteinHonsi - how exactly `ordinalSlope` and `ordinalOffset` work? Is this kind of an approximation function (ax + b)?

**Comment by TorsteinHonsi** [comment]
> Is this kind of an approximation function (ax + b)?

Yes, that's the idea. I don't have it all in my head, but I _think_ it applies to the extremes of the dataset.

**Pull request merged**